### PR TITLE
Fix image parsing in subctl show versions

### PR DESCRIPTION
**Pull request by maayanf24**
and rename IMAGE -> REPOSITORY in the title line

Signed-off-by: Maayan Friedman <maafried@redhat.com>

### CODE REVIEW

File: pkg/subctl/cmd/show_versions.go [hunk 1]
```diff
@@ -58,17 +58,9 @@ func getOperatorVersion(clientSet kubernetes.Interface, versions []versionImageI
...
-	img := strings.Split(operatorFullImageStr, ":")
-	var image string
-	var version string
-	if len(img) < 2 {
-		image = img[0]
-	} else {
-		image = img[0]
-		version = img[1]
-	}
+	i := strings.LastIndex(operatorFullImageStr, ":")
```
Comment (tpantelis) [inline pkg/subctl/cmd/show_versions.go:61]: If ':' isn't found, `i` will be -1 which causes a panic. This **shouldn't** happen but wouldn't hurt to handle as it did before.

**Comment by maayanf24** [comment]
> This still produces surprising results:
> 
> ```
> COMPONENT                       REPOSITORY                                      VERSION         
> submariner                      localhost:5000                                  local
> submariner-operator             localhost:5000/submariner-operator              local
> ```

@skitt  Can you be more specific?

**Comment by skitt** [comment]
> > This still produces surprising results:
> > ```
> > COMPONENT                       REPOSITORY                                      VERSION         
> > submariner                      localhost:5000                                  local
> > submariner-operator             localhost:5000/submariner-operator              local
> > ```
> 
> @skitt Can you be more specific?

I would expect

```
COMPONENT                       REPOSITORY                                      VERSION
submariner                      localhost:5000                                  local
submariner-operator             localhost:5000                                  local
```

or am I missing something?

**Comment by maayanf24** [comment]
> > > This still produces surprising results:
> > > ```
> > > COMPONENT                       REPOSITORY                                      VERSION         
> > > submariner                      localhost:5000                                  local
> > > submariner-operator             localhost:5000/submariner-operator              local
> > > ```
> > 
> > 
> > @skitt Can you be more specific?
> 
> I would expect
> 
> ```
> COMPONENT                       REPOSITORY                                      VERSION
> submariner                      localhost:5000                                  local
> submariner-operator             localhost:5000                                  local
> ```
> 
> or am I missing something?

for this case:
```
COMPONENT                       IMAGE                                           VERSION         
submariner                      localhost:5000                                  local           
submariner-operator             quay.io/submariner/submariner-operator          0.4.0         
```
we would like to keep the **submariner-operator** part, so if I'll remove it from the first case only it will be too specific parsing

**Comment by skitt** [comment]
> > > > This still produces surprising results:
> > > > ```
> > > > COMPONENT                       REPOSITORY                                      VERSION         
> > > > submariner                      localhost:5000                                  local
> > > > submariner-operator             localhost:5000/submariner-operator              local
> > > > ```
> > > 
> > > 
> > > @skitt Can you be more specific?
> > 
> > 
> > I would expect
> > ```
> > COMPONENT                       REPOSITORY                                      VERSION
> > submariner                      localhost:5000                                  local
> > submariner-operator             localhost:5000                                  local
> > ```
> > 
> > 
> > or am I missing something?
> 
> for this case:
> 
> ```
> COMPONENT                       IMAGE                                           VERSION         
> submariner                      localhost:5000                                  local           
> submariner-operator             quay.io/submariner/submariner-operator          0.4.0         
> ```
> 
> we would like to keep the **submariner-operator** part, so if I'll remove it from the first case only it will be too specific parsing

Ah right, I get it, thanks. Shouldn’t the local output be

```
COMPONENT                       REPOSITORY                                      VERSION         
submariner                      localhost:5000/submariner                       local
submariner-operator             localhost:5000/submariner-operator              local
```

in that case?

**Comment by maayanf24** [comment]
> Ah right, I get it, thanks. Shouldn’t the local output be
> 
> ```
> COMPONENT                       REPOSITORY                                      VERSION         
> submariner                      localhost:5000/submariner                       local
> submariner-operator             localhost:5000/submariner-operator              local
> ```
> 
> in that case?

yes. fixed

**Comment by mangelajo** [comment]
I think for local:
```
COMPONENT                       REPOSITORY                                      VERSION         
submariner                      localhost:5000                                     local
submariner-operator             localhost:5000                               local
```

For remote:

```
COMPONENT                       REPOSITORY                                      VERSION         
submariner                      quay.io/submariner                                    local
submariner-operator             quay.io/submariner                             local
```

The complicated part is the parsing on the operator image...

In the operator we will have different cases:
```

localhost:5000/submariner-operator                                -> localhost:5000
some-other-registry.com:1235/submariner-org/submariner-operator   -> some-other-registry.com:1235/submariner-org
submariner-org/submariner-operator                                -> submariner-org     (I think this would be valid for dockerhub, but not sure)
quay.io/submariner/submariner-operator                            -> quay.io/submariner

```

**Comment by tpantelis** [comment]
Are targeting this for 0.5.0?

**Comment by nyechiel** [comment]
> Are targeting this for 0.5.0?

No, targeting v0.6 now.

**Comment by maayanf24** [comment]
> [#547 (comment)](https://github.com/submariner-io/submariner-operator/pull/547#issuecomment-656014657)
> 
> See my comments here, we have probably some edge cases.
> 
> May be it's a good opportunity to add unit testing for some parts.

@mangelajo - Image parsing fixed according to your comment.
And I've added unit testing for the operator image parsing.

File: pkg/subctl/cmd/show_version_test.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	{"some-other-registry.com:1235/submariner-org/submariner-operator:v0.5.0", "some-other-registry.com:1235/submariner-org", "v0.5.0"},
+	{"submariner-org/submariner-operator:v0.4.0", "submariner-org", "v0.4.0"},
+	{"quay.io/submariner/submariner-operator:local", "quay.io/submariner", "local"},
+}
+
+func TestParseOperatorImage(t *testing.T) {
+	for _, tt := range imageTests {
+		version, repository := parseOperatorImage(tt.image)
+		if strings.Compare(repository, tt.repository) != 0 {
+			t.Fatalf("Operator repository for image %q shold be %q, but value parsed was %q\n", tt.image, tt.repository, repository)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/show_version_test.go:39]:
```suggestion
			t.Fatalf("Operator repository for image %q should be %q, but value parsed was %q\n", tt.image, tt.repository, repository)
```

File: pkg/subctl/cmd/show_version_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	for _, tt := range imageTests {
+		version, repository := parseOperatorImage(tt.image)
+		if strings.Compare(repository, tt.repository) != 0 {
+			t.Fatalf("Operator repository for image %q should be %q, but value parsed was %q\n", tt.image, tt.repository, repository)
+		}
+		if strings.Compare(version, tt.version) != 0 {
+			t.Fatalf("Operator version for image %q should be %q, but value parsed was %q\n", tt.image, tt.version, version)
+		}
+	}
+}
```
Comment (tpantelis) [inline pkg/subctl/cmd/show_version_test.go:45]: Please use the Gingko framework as we've standardized on that.

**Comment by tpantelis** [comment]
Need to add Signed-Off-By to the commits.

**Review by mangelajo — APPROVED** [review]
looks great now

**Pull request merged**